### PR TITLE
[Refactoring Tensor PR #4] remove the param-less data interface

### PR DESCRIPTION
**Pull request by Shixiaowei02**
### PR types

Others

### PR changes

Others

### Describe

为使代码更容易维护，为 Allocation 统一做基础，本提交处理了 Tensor Fluid 推全冗余的 `mutable_data` 接口。

### CODE REVIEW

File: paddle/pten/tests/api/test_conj_api.cc [hunk 1]
```diff
@@ -37,7 +37,8 @@ TEST(API, conj) {
       pten::DenseTensorMeta(pten::DataType::COMPLEX64,
                             framework::make_ddim({3, 10}),
                             pten::DataLayout::NCHW));
-  auto* dense_x_data = dense_x->mutable_data<paddle::complex64>();
+  auto* dense_x_data =
+      dense_x->mutable_data<paddle::complex64>(paddle::platform::CPUPlace());
```
Comment (chenwhql) [inline paddle/pten/tests/api/test_conj_api.cc:41]: place已经统一了，在pten中为什么不直接使用pten::CPUPlace

File: paddle/pten/core/dense_tensor.h [hunk 2]
```diff
@@ -172,24 +172,6 @@ class DenseTensor : public TensorBase,
   /// \return The actual storage size occupied by tensor.
   size_t capacity() const { return storage_->size(); }
 
-  /// \brief Get the mutable data pointer value of type T.
-  /// Memory allocation may occur when calling this interface:
-  /// 1. When the storage size is not enough to meet the current shape of the
-  /// data.
-  /// \return The mutable data pointer value of type T.
-  template <typename T>
-  T* mutable_data();
```
Comment (chenwhql) [inline paddle/pten/core/dense_tensor.h:181]: 为什么要删除这个接口，这个是明确会保留的

File: paddle/pten/kernels/impl/matmul_kernel_impl.h [hunk 3]
```diff
@@ -176,7 +176,7 @@ void MatMulFunction(const Context& dev_ctx,
                 y_data,
                 x_data,
                 static_cast<T>(flag),
-                Out->mutable_data<T>());
+                Out->data<T>());
```
Comment (jim19930609) [inline paddle/pten/kernels/impl/matmul_kernel_impl.h:179]: Replace with "mutable_data<T>(dev_ctx.GetPlace())"

File: paddle/pten/kernels/impl/matmul_kernel_impl.h [hunk 4]
```diff
@@ -189,7 +189,7 @@ void MatMulFunction(const Context& dev_ctx,
                   y_data,
                   x_data,
                   static_cast<T>(flag),
-                  Out->mutable_data<T>());
+                  Out->data<T>());
```
Comment (jim19930609) [inline paddle/pten/kernels/impl/matmul_kernel_impl.h:192]: ditto

File: paddle/pten/kernels/impl/matmul_kernel_impl.h [hunk 5]
```diff
@@ -201,7 +201,7 @@ void MatMulFunction(const Context& dev_ctx,
                          y_data,
                          x_data,
                          static_cast<T>(flag),
-                         Out->mutable_data<T>(),
+                         Out->data<T>(),
```
Comment (jim19930609) [inline paddle/pten/kernels/impl/matmul_kernel_impl.h:204]: ditto

File: paddle/pten/kernels/impl/matmul_kernel_impl.h [hunk 6]
```diff
@@ -257,7 +257,7 @@ void MatMulFunction(const Context& dev_ctx,
                   x_data,
                   y_data,
                   static_cast<T>(flag),
-                  Out->mutable_data<T>());
+                  Out->data<T>());
```
Comment (jim19930609) [inline paddle/pten/kernels/impl/matmul_kernel_impl.h:260]: ditto

File: paddle/pten/kernels/impl/matmul_kernel_impl.h [hunk 7]
```diff
@@ -269,7 +269,7 @@ void MatMulFunction(const Context& dev_ctx,
                          x_data,
                          y_data,
                          static_cast<T>(flag),
-                         Out->mutable_data<T>(),
+                         Out->data<T>(),
```
Comment (jim19930609) [inline paddle/pten/kernels/impl/matmul_kernel_impl.h:272]: ditto

File: paddle/pten/kernels/impl/matmul_kernel_impl.h [hunk 8]
```diff
@@ -284,7 +284,7 @@ void MatMulFunction(const Context& dev_ctx,
                 x_data,
                 y_data,
                 static_cast<T>(flag),
-                Out->mutable_data<T>());
+                Out->data<T>());
```
Comment (jim19930609) [inline paddle/pten/kernels/impl/matmul_kernel_impl.h:287]: ditto

File: paddle/pten/kernels/impl/matmul_kernel_impl.h [hunk 9]
```diff
@@ -367,7 +367,7 @@ void MatMulFunction(const Context& dev_ctx,
               x_data,
               y_data,
               static_cast<T>(flag),
-              Out->mutable_data<T>());
+              Out->data<T>());
```
Comment (jim19930609) [inline paddle/pten/kernels/impl/matmul_kernel_impl.h:370]: ditto

File: paddle/pten/kernels/impl/matmul_grad_kernel_impl.h [hunk 10]
```diff
@@ -123,7 +123,7 @@ void MatMul(const Context& dev_ctx,
               b.data<T>(),
               mat_dim_b,
               static_cast<T>(1),
-              out->mutable_data<T>(),
+              out->data<T>(),
```
Comment (jim19930609) [inline paddle/pten/kernels/impl/matmul_grad_kernel_impl.h:126]: replace with "mutable_data<T>(dev_ctx.GetPlace())"

File: paddle/pten/core/dense_tensor.cc [hunk 11]
```diff
@@ -439,8 +387,7 @@ template <typename T>
 inline T* DenseTensor::mutable_data(const paddle::platform::Place& place,
                                     size_t requested_size) {
   static_assert(std::is_pod<T>::value, "T must be POD");
-  return reinterpret_cast<T*>(mutable_data(
-      place, paddle::framework::DataTypeTrait<T>::DataType(), requested_size));
+  return reinterpret_cast<T*>(mutable_data(place, requested_size));
```
Comment (jim19930609) [inline paddle/pten/core/dense_tensor.cc]: infinite recursive call?

File: paddle/pten/tests/api/test_conj_api.cc (hunk 1 above)
Comment (Shixiaowei02) [reply paddle/pten/tests/api/test_conj_api.cc:41]: 好的。看了一下，pten 还没统一替换，这部分全局文本替换比较简单，后续我们统一提交吧

File: paddle/pten/core/dense_tensor.h (hunk 2 above)
Comment (Shixiaowei02) [reply paddle/pten/core/dense_tensor.h:181]: 因为与已有 `template T* DenseTensor::data()` 重复了，所以为了推进后续工作，先删除

File: paddle/pten/core/dense_tensor.cc (hunk 11 above)
Comment (Shixiaowei02) [reply paddle/pten/core/dense_tensor.cc]: I thought it was a non-template function called. I will confirm, thanks

File: paddle/pten/kernels/impl/matmul_kernel_impl.h (hunk 3 above)
Comment (Shixiaowei02) [reply paddle/pten/kernels/impl/matmul_kernel_impl.h:179]: I found out that the assignment has been performed before.

File: paddle/pten/kernels/impl/matmul_kernel_impl.h (hunk 4 above)
Comment (Shixiaowei02) [reply paddle/pten/kernels/impl/matmul_kernel_impl.h:192]: +1, thanks!

File: paddle/pten/kernels/impl/matmul_kernel_impl.h (hunk 5 above)
Comment (Shixiaowei02) [reply paddle/pten/kernels/impl/matmul_kernel_impl.h:204]: +1, thanks!

File: paddle/pten/kernels/impl/matmul_kernel_impl.h (hunk 6 above)
Comment (Shixiaowei02) [reply paddle/pten/kernels/impl/matmul_kernel_impl.h:260]: +1, thanks!

File: paddle/pten/kernels/impl/matmul_kernel_impl.h (hunk 7 above)
Comment (Shixiaowei02) [reply paddle/pten/kernels/impl/matmul_kernel_impl.h:272]: +1, thanks!

File: paddle/pten/kernels/impl/matmul_kernel_impl.h (hunk 8 above)
Comment (Shixiaowei02) [reply paddle/pten/kernels/impl/matmul_kernel_impl.h:287]: +1, thanks!

File: paddle/pten/kernels/impl/matmul_kernel_impl.h (hunk 9 above)
Comment (Shixiaowei02) [reply paddle/pten/kernels/impl/matmul_kernel_impl.h:370]: +1, thanks!

File: paddle/pten/kernels/impl/matmul_grad_kernel_impl.h (hunk 10 above)
Comment (Shixiaowei02) [reply paddle/pten/kernels/impl/matmul_grad_kernel_impl.h:126]: +1, Thanks for your review!

**Review by jiweibo — DISMISSED** [review]
LGTM

**Review by chenwhql — DISMISSED** [review]

**Pull request closed without merging**